### PR TITLE
SubscriptionTopicDispatcher

### PR DESCRIPTION
**Pull request by fil512**
Create a new SubscriptionTopicDispatcher service for use by java extensions that need to dispatch their own subscription topic notifications

### CODE REVIEW

**Review by jmarchionatto — COMMENTED** [review]
Posted a question

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicDispatcher.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	 */
+	public void dispatch(@Nonnull String theTopicUrl, @Nonnull List<IBaseResource> theResources, RestOperationTypeEnum theRequestType, @Nullable InMemoryMatchResult theInMemoryMatchResult) {
+		List<ActiveSubscription> topicSubscriptions = mySubscriptionRegistry.getTopicSubscriptionsByTopic(theTopicUrl);
+		if (!topicSubscriptions.isEmpty()) {
+			for (ActiveSubscription activeSubscription : topicSubscriptions) {
+				// WIP STR5 apply subscription filters
+				IBaseBundle bundlePayload = mySubscriptionTopicPayloadBuilder.buildPayload(theResources, activeSubscription, theTopicUrl, theRequestType);
+				// WIP STR5 do we need to add a total?  If so can do that with R5BundleFactory
+				bundlePayload.setId(UUID.randomUUID().toString());
+				ResourceModifiedMessage resourceModifiedMessage = new ResourceModifiedMessage();
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicDispatcher.java:49]: Oh shoot.  You're right.  There is a serious error here.  Good catch!

[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicDispatcher.java]: fixed

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicPayloadBuilder.java [hunk 2]
```diff
@@ -65,31 +65,42 @@ public IBaseBundle buildPayload(IBaseResource theMatchedResource, ResourceModifi
...
 		subscriptionStatus.setId(UUID.randomUUID().toString());
 		subscriptionStatus.setStatus(Enumerations.SubscriptionStatusCodes.ACTIVE);
 		subscriptionStatus.setType(SubscriptionStatus.SubscriptionNotificationType.EVENTNOTIFICATION);
 		// WIP STR5 count events since subscription start and set eventsSinceSubscriptionStart
 		// store counts by subscription id
 		subscriptionStatus.setEventsSinceSubscriptionStart(theEventsSinceSubscriptionStart);
-		subscriptionStatus.addNotificationEvent().setEventNumber(theEventsSinceSubscriptionStart).setFocus(new Reference(theMatchedResource.getIdElement()));
+		SubscriptionStatus.SubscriptionStatusNotificationEventComponent event = subscriptionStatus.addNotificationEvent();
+		event.setEventNumber(theEventsSinceSubscriptionStart);
+		if (theResources.size() > 0) {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicPayloadBuilder.java:98]: I don't understand.  In this case, there is no resource to focus on.

**Review by jmarchionatto — APPROVED** [review]
Looks good. Approved with a couple of minor comments

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionMatchDeliverer.java [hunk 3]
```diff
@@ -51,41 +53,69 @@ public SubscriptionMatchDeliverer(FhirContext theFhirContext, IInterceptorBroadc
...
+		SubscriptionDeliveryRequest subscriptionDeliveryRequest;
+		if (thePayload != null) {
+			subscriptionDeliveryRequest = new SubscriptionDeliveryRequest(thePayload, theMsg, theActiveSubscription);
+		} else {
+			subscriptionDeliveryRequest = new SubscriptionDeliveryRequest(theMsg.getPayloadId(myFhirContext), theMsg, theActiveSubscription);
+		}
+		ResourceDeliveryMessage deliveryMsg = buildResourceDeliveryMessage(subscriptionDeliveryRequest);
+		deliveryMsg.copyAdditionalPropertiesFrom(theMsg);
+
+		if (!callHooks(theActiveSubscription, theInMemoryMatchResult, deliveryMsg)) {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionMatchDeliverer.java:66]: We don't usually do that elsewhere, so I think we can leave it to the hook to log that

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionMatchDeliverer.java [hunk 4]
```diff
@@ -51,41 +53,69 @@ public SubscriptionMatchDeliverer(FhirContext theFhirContext, IInterceptorBroadc
...
-		String subscriptionId = theActiveSubscription.getId();;
+	public boolean deliverPayload(@Nonnull SubscriptionDeliveryRequest subscriptionDeliveryRequest, InMemoryMatchResult theInMemoryMatchResult) {
+		ResourceDeliveryMessage deliveryMsg = buildResourceDeliveryMessage(subscriptionDeliveryRequest);
+
+		if (!callHooks(subscriptionDeliveryRequest.getActiveSubscription(), theInMemoryMatchResult, deliveryMsg)) {
+			return false;
+		}
+
+		return sendToDeliveryChannel(subscriptionDeliveryRequest.getActiveSubscription(), deliveryMsg);
+	}
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionMatchDeliverer.java]: fixed

**Pull request merged**